raiden_contracts/contract_manager.py

```diff
@@ -85,77 +91,99 @@ def _compile_all_contracts(self) -> None:
+                        overall_checksum=self.overall_checksum,
+                    ),
+                ),
             )
 
     def get_contract(self, contract_name: str) -> Dict:
         """ Return ABI, BIN of the given contract. """
-        if not self._contracts:
+        if not self.contracts:
             self._compile_all_contracts()
-        return self._contracts[contract_name]
+        return self.contracts[contract_name]
 
     def get_contract_abi(self, contract_name: str) -> Dict:
         """ Returns the ABI for a given contract. """
-        if not self._contracts:
+        if not self.contracts:
             self._compile_all_contracts()
-        return self._contracts[contract_name]['abi']
+        return self.contracts[contract_name]['abi']
 
     def get_event_abi(self, contract_name: str, event_name: str) -> Dict:
         """ Returns the ABI for a given event. """
         # Import locally to avoid web3 dependency during installation via `compile_contracts`
         from web3.utils.contracts import find_matching_event_abi
 
-        if not self._contracts:
+        if not self.contracts:
             self._compile_all_contracts()
         contract_abi = self.get_contract_abi(contract_name)
         return find_matching_event_abi(contract_abi, event_name)
 
-    def _checksum_contracts(self):
-        if self._contracts_source_dirs is None:
-            raise TypeError("Can't checksum when using precompiled contracts.")
-        checksums = []
-        for contracts_dir in self._contracts_source_dirs.values():
+    def checksum_contracts(self) -> None:
+        if self.contracts_source_dirs is None:
+            raise TypeError("Missing contracts source path, can't checksum contracts.")
+
+        checksums = {}
+        for contracts_dir in self.contracts_source_dirs.values():
             file: Path
             for file in contracts_dir.glob('*.sol'):
-                checksums.append(
-                    '{}:{}'.format(
-                        file.name,
-                        hashlib.sha256(file.read_bytes()).hexdigest(),
-                    ),
+                checksums[file.name] = hashlib.sha256(file.read_bytes()).hexdigest()
+
+        self.overall_checksum = hashlib.sha256(
+            ':'.join([checksums[key] for key in sorted(checksums)]).encode(),
+        ).hexdigest()
+        self.contracts_checksums = checksums
+
+    def verify_precompiled_checksums(self, contracts_precompiled: str) -> None:
+        """ Compare source code checksums with those from a precompiled file. """
+
+        # We get the precompiled file data
+        contracts_precompiled = ContractManager(contracts_precompiled)
+
+        # Compare each contract source code checksum with the one from the precompiled file
+        for contract, checksum in self.contracts_checksums.items():
+            try:
+                precompiled_checksum = contracts_precompiled.contracts_checksums[contract]
```

Comment:
To avoid a regression of what you mentioned in Rocketchat, perhaps add an assertion for the contract names here? They are also kept in the contracts.json file so you can also check for them here, right?

Reply:
how would an assertion here avoid that regression?

Reply:
Hmm correct. I did not word that correctly.

It would crash, but with a nice message explaining what it was expecting and we could immediately deduce what we did wrong.

Reply:
This would not fix the problem. There was no key missing in the dictionary, the problem is that the order of the keys is not stable. The existing code `hashlib.sha256(':'.join(checksums).encode()).hexdigest()` was not stable if the type of checksums is a dictionary.

Reply:
Side note: this made me ask myself if we also check if all the contracts were compiled and checksummed.
We check that they were all checksummed by `overall_checksum`, which would fail if a contract was missing.
Should I also check here if `contracts.json[contracts]` contains the compiled info for all contracts?  E.g. `contracts_precompiled.contracts[contract] is not None` ?

Reply:
Opened https://github.com/raiden-network/raiden-contracts/issues/301 to see if there are other things to be checked for, so you can comment there. I wouldn't expand this PR more..